syntax = "proto3";

import "google/protobuf/empty.proto";

option csharp_namespace = "DataProviderService.Protos";

package data_provider_service.v1;

service StateStore {
  rpc ReadStates (ReadStatesRequest) returns (StateValues);
  rpc WriteStates (WriteStatesRequest) returns (google.protobuf.Empty);
  rpc SubscribeStates (SubscribeStatesRequest) returns (stream StateStreamValues);
  rpc SubscribeStateChangeEvents (SubscribeStateChangeEventsRequest) returns (stream StateEvent);
}

message ReadStatesRequest {
  repeated uint64 state_ids = 1;
  int64 updated_later_than = 2;
}

message WriteStatesRequest {
  repeated StateValue values = 1;
}

message SubscribeStatesRequest {
  repeated uint64 state_ids = 1;
  bool notify_empty_change_sets = 2;
  bool notify_unchanged = 3;
  uint32 minimum_notification_interval_ms = 4;
}

message StateValues {
  sint64 read_time = 1;
  repeated StateValue values = 2;
}

message StateStreamValues {
  repeated StateValue values = 1;
}

message SubscribeStateChangeEventsRequest {
  repeated uint64 state_ids = 1;
}

message StateEvent {
  uint64 state_id = 1;
  oneof change {
    BooleanOldNew boolean_change = 2;
	Int32OldNew int32_change = 3;
	DoubleOldNew double_change = 4;
    StringOldNew string_change = 5;
	Double2OldNew double2_change = 6;
	Double3OldNew double3_change = 7;
	Double4OldNew double4_change = 8;
	Double5OldNew double5_change = 9;
  }
}

message StateValue {
  uint64 state_id = 1;
  oneof value {
	bool boolean_value = 2;
	sint32 int32_value = 3;
	double double_value = 4;
	string string_value = 5;
	Double2 double2_value = 6;
	Double3 double3_value = 7;
	Double4 double4_value = 8;
	Double5 double5_value = 9;
  }
}

message Double2 {
  double value_0 = 1;
  double value_1 = 2;
}

message Double3 {
  double value_0 = 1;
  double value_1 = 2;
  double value_2 = 3;
}

message Double4 {
  double value_0 = 1;
  double value_1 = 2;
  double value_2 = 3;
  double value_3 = 4;
}

message Double5 {
  double value_0 = 1;
  double value_1 = 2;
  double value_2 = 3;
  double value_3 = 4;
  double value_4 = 5;
}

message BooleanOldNew {
  bool old_value = 1;
  bool new_value = 2;
}

message Int32OldNew {
  sint32 old_value = 1;
  sint32 new_value = 2;
}

message DoubleOldNew {
  double old_value = 1;
  double new_value = 2;
}

message StringOldNew {
  string old_value = 1;
  string new_value = 2;
}

message Double2OldNew {
  Double2 old_value = 1;
  Double2 new_value = 2;
}

message Double3OldNew {
  Double3 old_value = 1;
  Double3 new_value = 2;
}

message Double4OldNew {
  Double4 old_value = 1;
  Double4 new_value = 2;
}

message Double5OldNew {
  Double5 old_value = 1;
  Double5 new_value = 2;
}
